src/server/main_service.cc

```diff
@@ -190,28 +190,33 @@ auto CmdEntryToMonitorFormat(std::string_view str) -> std::string {
   return result;
 }
 
-std::string MakeMonitorMessage(const ConnectionState& conn_state,
-                               const facade::Connection* connection, CmdArgList args) {
-  std::string message = absl::StrCat(CreateMonitorTimestamp(), " [", conn_state.db_index);
-
-  if (conn_state.script_info) {
-    absl::StrAppend(&message, " lua] ");
-  } else {
-    auto endpoint = connection == nullptr ? "REPLICATION:0" : connection->RemoteEndpointStr();
-    absl::StrAppend(&message, " ", endpoint, "] ");
-  }
-  if (args.empty()) {
-    absl::StrAppend(&message, "error - empty cmd list!");
-  } else if (auto cmd_name = std::string_view(args[0].data(), args[0].size());
-             cmd_name == "AUTH") {  // we cannot just send auth details in this case
-    absl::StrAppend(&message, "\"", cmd_name, "\"");
+std::string MakeMonitorMessage(const ConnectionContext* cntx, const CommandId* cid,
+                               CmdArgList tail_args) {
+  std::string message = absl::StrCat(CreateMonitorTimestamp(), " [", cntx->conn_state.db_index);
+
+  if (cntx->conn_state.squashing_info)
+    cntx = cntx->conn_state.squashing_info->owner;
+
+  string endpoint;
+  if (cntx->conn_state.script_info) {
+    endpoint = "lua";
+  } else if (const auto* conn = cntx->owner(); conn != nullptr) {
+    endpoint = conn->RemoteEndpointStr();
   } else {
-    message = std::accumulate(args.begin(), args.end(), message, [](auto str, const auto& cmd) {
-      absl::StrAppend(&str, " ", CmdEntryToMonitorFormat(std::string_view(cmd.data(), cmd.size())));
-      return str;
-    });
+    endpoint = "REPLICATION:0";
   }
-  return message;
+  absl::StrAppend(&message, " ", endpoint, "] ");
+
+  absl::StrAppend(&message, "\"", cid->name(), "\"");
+
+  if (cid->name() == "AUTH")
+    return message;
+
+  auto accum = [](auto str, const auto& cmd) {
```

Comment:
`accumulate` is a really nice idea here. However it suffers from two issues:

1. `auto str` is a deep copy each time the lambda is invoked by the accumulate algorithm. I think only in c++20 the `init` string `message` is moved, by a call to `std::move`. So basically now, we are copying `tail_args.size()` strings with each of them increasing in size at each step (since we append to the string at each step).
2. Furthermore, another copy is done when we return `str`. This is not one of the cases were NRVO applies, because `str` is an input parameter of the lambda.

All in all, I would suggest to rewrite this with a for loop or a for-each.

Reply:
True, its code from about a year ago that I just copied over 😄 (didn't think about optimizing because monitor is not critical)